File: 1248-count-number-of-nice-subarrays/1248-count-number-of-nice-subarrays.java
class Solution {
    public int numberOfSubarrays(int[] nums, int k) {
       int n = nums.length;
        int[] prefix = new int[n + 1];

        //detect odd and even
        for (int i = 0; i < n; i++) {
            nums[i] = (nums[i] % 2)==0?0:1;
        }

        //add prefix
        for (int i = 1; i <=n; i++) {
            prefix[i] = prefix[i-1] + nums[i-1];
        }

        int res=0;
        Map<Integer, Integer> sumPrefixMap = new HashMap<>();
         sumPrefixMap.put(0, 1);
        for (int i = 1; i <= n; i++) {
             sumPrefixMap.merge(prefix[i], 1, Integer::sum);
             if (sumPrefixMap.containsKey(prefix[i] - k)) {
                    res += sumPrefixMap.get(prefix[i] - k);
             }
        }


        return res;
    }
}

File: 1248-count-number-of-nice-subarrays/1248-count-number-of-nice-subarrays.py
class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
        
        pref=list(accumulate([0 if num&1==0 else 1 for num in nums]))
        mp,ans=defaultdict(int),0
        
        mp[0]=1
        for num in pref:
            if num-k in mp:
                ans+=mp[num-k]
            mp[num]+=1
        return ans
                
                
            